Appendix
########

Literature
==========

Algorithms
----------

* Bowtie:
  Langmead B, Trapnell C, Pop M, Salzberg SL. “Ultrafast and memory-efficient alignment of short DNA sequences to the human genome”, Genome Biology <http://genomebiology.com/2009/10/3/R25>. 

* Tophat:
  Trapnell C, Pachter L, Salzberg SL, TopHat: discovering splice junctions with RNA-Seq, Bioinformatics, 25(9):1105-1111 <http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btp120>. 

* Cufflinks:
  Trapnell C, Williams BA, Pertea G, Mortazavi AM, Kwan G, van Baren MJ, Salzberg SL, Wold B, Pachter L.Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation <http://dx.doi.org/10.1038/nbt.1621> Nature Biotechnology

* BWA: 
  Li et al " Fast and accurate short read alignment with Burrows-Wheeler transform" http://bioinformatics.oxfordjournals.org/content/25/14/1754.abstract

File Manipulation tools
-----------------------

* Samtools - http://samtools.sourceforge.net/

* Picard - http://picard.sourceforge.net/index.shtml

* FastQC - http://www.bioinformatics.babraham.ac.uk/projects/fastqc/

* HTSeq - http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html


Data Standards
--------------

* http://encodeproject.org/ENCODE/dataStandards.html

* Sam format - http://samtools.sourceforge.net/SAM1.pdf

* http://www.ebi.ac.uk/arrayexpress/

* http://www.ebi.ac.uk/ena/


Normalization
--------------

* Oshlack A. et al "From RNA-seq reads to differential expression results." Genome Biology 

* Robinson and Oshlack "A scaling normalization method for differential expression analysis of RNA-seq data." Genome Biology 

* http://www.biomedcentral.com/1471-2105/12/480/abstract - GC content normalization

* http://www.bepress.com/ucbbiostat/paper291/ - GC bias

Differential Expression
-----------------------

* Tarazona et al. Differential expression in RNA-seq: A matter of depth, 2011,
  Genome research 21 (12) p. 2213-23
  <http://genome.cshlp.org/cgi/content/abstract/gr.124321.111v1>

* Anders and Huber, "Differential expression analysis for sequence count data", 2010, Genome Biology, 11 (10) p. R106
  <http://genomebiology.com/2010/11/10/R106>

* Robinson et al, "edgeR: a Bioconductor package for differential expression analysis of digital gene expression data", 2010, Bioinformatics (Oxford, England) 26 (1) p. 139-40
  < http://bioinformatics.oxfordjournals.org/content/26/1/139.long>



