<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Aligning RNASeq Reads</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="SAMtools" href="STEP3_viewManipulateAlignmentData_samtools.html" />
    <link rel="prev" title="Quality Control" href="STEP1_QualityControl.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="STEP3_viewManipulateAlignmentData_samtools.html" title="SAMtools"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="STEP1_QualityControl.html" title="Quality Control"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">None</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="aligning-rnaseq-reads">
<span id="tophat"></span><h1>Aligning RNASeq Reads<a class="headerlink" href="#aligning-rnaseq-reads" title="Permalink to this headline">¶</a></h1>
<div class="section" id="example-1-arm-xsubset-fastq">
<h2>Example 1 - arm.Xsubset.fastq<a class="headerlink" href="#example-1-arm-xsubset-fastq" title="Permalink to this headline">¶</a></h2>
<p>The dataset <tt class="docutils literal"><span class="pre">arm.Xsubset.fastq</span></tt> contains reads from an mRNA sample after RNAi knockdown of the <a class="reference external" href="http://flybase.org/reports/FBgn0000117.html">arm</a> gene, in the <em>Drosophila melanogaster</em> cell line, S2. The <em>arm</em> gene is found on chromosome X of the <em>Drosophila melanogaster</em> genome. Here, to save time, we will restrict our analysis to this chromosome. For your own projects, you should use all of the genome information.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">In your projects you need to align to the full genome of your organism of interest, since you don&#8217;t know which chromosome the reads come from. It might be the case that the reads match better elsewhere on the genome.</p>
</div>
<p>Your task is to align the reads to the X chromosome. To align reads to the X chromosome, we can use a command line tool called <strong>Tophat</strong>.</p>
<p>To get information on <strong>Tophat</strong>, open a terminal window and type:</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; tophat -h
</pre></div>
</div>
</div></blockquote>
<p>You will find a file called <tt class="docutils literal"><span class="pre">arm.Xsubset.fastq</span></tt> in a folder called <tt class="docutils literal"><span class="pre">rawData</span></tt> in the course data folder. The data in the <tt class="docutils literal"><span class="pre">arm.Xsubset.fastq</span></tt> file is the raw data that you need to align to chromosome X.</p>
<p>The annotation data for chromosome X is in the file <tt class="docutils literal"><span class="pre">geneRef/drosophilaMelanogaster.X.gtf</span></tt></p>
<p>The bowtie index is <tt class="docutils literal"><span class="pre">bowtieIndexes/drosophilaMelanogaster.X</span></tt></p>
<p>Create an output directory where we can store the results of the alignment. Remember that you can create a new directory with the following command (see section <a class="reference internal" href="../introduction.html#mkdir"><em>mkdir – Create a new directory</em></a>).</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; mkdir -p ~/arm_tophatOutput
</pre></div>
</div>
</div></blockquote>
<p>We can use the following command to run <strong>Tophat</strong>, to perform an alignment of the NGS reads to the annotation data for chromosome X:</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; tophat --no-coverage-search -G geneRef/drosophila.2Lsubset.gtf -o ~/arm_tophatOutput/ <span class="se">\</span>
  bowtieIndexes/drosophilaMelanogaster.X rawData/arm.Xsubset.fastq
</pre></div>
</div>
</div></blockquote>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Remember to specify the output directory in the tophat command.</p>
</div>
<div class="topic">
<p class="topic-title first">Review questions</p>
<ul class="simple">
<li>What does <strong>Tophat</strong> do?</li>
<li>What kind of aligner is <strong>Tophat</strong>?</li>
<li>What is the <tt class="docutils literal"><span class="pre">bowtie-index</span></tt>?</li>
<li>How can specifying more mismatches with the <tt class="docutils literal"><span class="pre">-n</span></tt> option change the output?</li>
<li>Why supply a gtf file to <strong>Tophat</strong>?</li>
<li>option <tt class="docutils literal"><span class="pre">--no-coverage-search</span></tt> in the <strong>Tophat</strong> command will speed things up. Why?</li>
<li>What are the output files from the <strong>Tophat</strong> aligner?</li>
<li>What is the gtf file and why do I need it?</li>
<li>What data does the gtf file contain?</li>
<li>What does the last column in the gtf file contain?</li>
</ul>
</div>
</div>
<div class="section" id="example-2-smo-2lsubset-fastq">
<h2>Example 2 - smo.2Lsubset.fastq<a class="headerlink" href="#example-2-smo-2lsubset-fastq" title="Permalink to this headline">¶</a></h2>
<p>The dataset <tt class="docutils literal"><span class="pre">smo.2Lsubset.fastq</span></tt> contains reads from a mRNA sample after RNAi knockdown of the <a class="reference external" href="http://flybase.org/reports/FBgn0003444.html">smo</a> gene, in the <em>Drosophila melanogaster</em> cell line, S2. The <em>smo</em> gene is found on chromosome 2L of the <em>Drosophila melanogaster</em> genome. Once again we will restrict our analysis to a single chromosome.</p>
<p>Your task is to align the reads from the to the 2L chromosome</p>
<p>The raw data is called <tt class="docutils literal"><span class="pre">smo.2Lsubset.fastq</span></tt> and can be found in the <tt class="docutils literal"><span class="pre">rawData</span></tt> directory.</p>
<p>The annotation data for chromosome 2L is: <tt class="docutils literal"><span class="pre">geneRef/drosophilaMelanogaster.2L.gtf</span></tt></p>
<p>The bowtie index is <tt class="docutils literal"><span class="pre">bowtieIndex/drosophilaMelanogaster.2L</span></tt></p>
<div class="section" id="task">
<h3>Task<a class="headerlink" href="#task" title="Permalink to this headline">¶</a></h3>
<p>Which commands need to be run to align the reads to the 2L chromosome?</p>
<div class="topic">
<p class="topic-title first">Review questions</p>
<ul class="simple">
<li>What does the <tt class="docutils literal"><span class="pre">-p</span></tt> option in tophat do?</li>
<li>How does <tt class="docutils literal"><span class="pre">--no-coverage</span></tt> search save time?</li>
<li>What does the <tt class="docutils literal"><span class="pre">-g</span></tt> option do?</li>
</ul>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/HUBLogo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Aligning RNASeq Reads</a><ul>
<li><a class="reference internal" href="#example-1-arm-xsubset-fastq">Example 1 - arm.Xsubset.fastq</a></li>
<li><a class="reference internal" href="#example-2-smo-2lsubset-fastq">Example 2 - smo.2Lsubset.fastq</a><ul>
<li><a class="reference internal" href="#task">Task</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="STEP1_QualityControl.html"
                        title="previous chapter">Quality Control</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="STEP3_viewManipulateAlignmentData_samtools.html"
                        title="next chapter">SAMtools</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/exercises/STEP2_alignRNASeqReads.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="STEP3_viewManipulateAlignmentData_samtools.html" title="SAMtools"
             >next</a> |</li>
        <li class="right" >
          <a href="STEP1_QualityControl.html" title="Quality Control"
             >previous</a> |</li>
        <li><a href="../index.html">None</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Grainne Kerr, Holger Dinkel, Jon Fuller, Matt Betts.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>