<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>SAMtools</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="Estimating Expression" href="STEP4_CUFFLINKS_DE.html" />
    <link rel="prev" title="Aligning RNASeq Reads" href="STEP2_alignRNASeqReads.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="STEP4_CUFFLINKS_DE.html" title="Estimating Expression"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="STEP2_alignRNASeqReads.html" title="Aligning RNASeq Reads"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">None</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="samtools">
<h1>SAMtools<a class="headerlink" href="#samtools" title="Permalink to this headline">¶</a></h1>
<p><strong>SAMtools</strong> is a set of scripts (a toolbox so to say) that can be used to manipulate and view sam/bam files. In particular you can: <tt class="docutils literal"><span class="pre">sort</span></tt>, <tt class="docutils literal"><span class="pre">index</span></tt>, <tt class="docutils literal"><span class="pre">merge</span></tt>, and <tt class="docutils literal"><span class="pre">view</span></tt> these files.</p>
<p>There is more information available on the <strong>SAMtools</strong> websites</p>
<ul class="simple">
<li><a class="reference external" href="http://samtools.sourceforge.net/">http://samtools.sourceforge.net/</a></li>
<li><a class="reference external" href="http://samtools.sourceforge.net/samtools-c.shtml">http://samtools.sourceforge.net/samtools-c.shtml</a></li>
<li><a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a></li>
</ul>
<p><strong>Getting help</strong></p>
<p>To print a list of all the tools available in the samtools suite, simply type <tt class="docutils literal"><span class="pre">samtools</span></tt> on the command line</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools
</pre></div>
</div>
</div></blockquote>
<p>To print a list of the parameters required and options available for each tool in the suite, simply type <tt class="docutils literal"><span class="pre">samtools</span></tt> followed by the name of the tool on the command line. For example, to get a list of the options available for the <tt class="docutils literal"><span class="pre">view</span></tt> tool in the samtools suite, simply type:</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools view
</pre></div>
</div>
</div></blockquote>
<div class="section" id="view-bam-files">
<h2>View bam files<a class="headerlink" href="#view-bam-files" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">smo.bam</span></tt> is a bam file and is not human readable. To make it human readable you can convert it to a sam file.</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools view -h -o ~/smo.sam rawData/smo.bam
</pre></div>
</div>
</div></blockquote>
<ul class="simple">
<li>What does <tt class="docutils literal"><span class="pre">-h</span></tt> and <tt class="docutils literal"><span class="pre">-o</span></tt> do in the above example?</li>
<li>Convert <tt class="docutils literal"><span class="pre">arm.bam</span></tt> into a sam file.</li>
<li>Use the <tt class="docutils literal"><span class="pre">samtools</span> <span class="pre">view</span></tt> (and read the help) to view a specific region e.g. all reads mapping to chromosome X</li>
</ul>
</div>
<div class="section" id="viewing-the-header-of-a-bam-file">
<h2>Viewing the header of a bam file<a class="headerlink" href="#viewing-the-header-of-a-bam-file" title="Permalink to this headline">¶</a></h2>
<p>In some cases you might only want to see or generate the header of a bam file.</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools view -H rawData/smo.bam
</pre></div>
</div>
</div></blockquote>
<p>Now try:</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools view -H rawData/smo.bam &gt; ~/smo.header.sam
</pre></div>
</div>
</div></blockquote>
<ul class="simple">
<li>What does the <tt class="docutils literal"><span class="pre">&gt;</span> <span class="pre">~/smo.header.sam</span></tt> of the above statement do?</li>
<li>What information is stored in the header of the sam file?</li>
<li>From the header of the file, can you tell which alignment program was used to generate the bam file</li>
</ul>
</div>
<div class="section" id="count-the-number-of-alignments-in-a-bam-file">
<h2>Count the number of alignments in a bam file<a class="headerlink" href="#count-the-number-of-alignments-in-a-bam-file" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools view -c rawData/smo.bam
</pre></div>
</div>
</div></blockquote>
<ul class="simple">
<li>Can you use samtools to count the number of alignments above a quality score of 20 in your file?</li>
<li>What does the quality score of an alignment indicate?</li>
<li>How many alignments are in the <tt class="docutils literal"><span class="pre">arm.bam</span></tt> above a quality score of 50</li>
<li>What other flags/filtering options are there?</li>
</ul>
</div>
<div class="section" id="create-a-bam-index">
<h2>Create a bam index<a class="headerlink" href="#create-a-bam-index" title="Permalink to this headline">¶</a></h2>
<p>Use <tt class="docutils literal"><span class="pre">samtools</span> <span class="pre">index</span></tt> to create an index of <tt class="docutils literal"><span class="pre">smo.bam</span></tt></p>
<p>Get help</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools index
</pre></div>
</div>
</div></blockquote>
<p>Note the usage of the index command in the samtools toolbox suite.</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools index rawData/smo.bam smo.bai
</pre></div>
</div>
</div></blockquote>
<ul class="simple">
<li>Does the above command work as written? If not, what do you need to change?</li>
<li>What does creating a bam index mean?</li>
<li>Why would one want to create a bam index?</li>
</ul>
</div>
<div class="section" id="sorting-the-sam-file">
<h2>Sorting the sam file<a class="headerlink" href="#sorting-the-sam-file" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools sort rawData/smo.sam
</pre></div>
</div>
</div></blockquote>
<ul class="simple">
<li>What does <tt class="docutils literal"><span class="pre">-o</span></tt> in the above command do?</li>
<li>Change the command to sort by read names rather than chromosomal locations.</li>
<li>Change the above command so that the sorted reads are outputted to <tt class="docutils literal"><span class="pre">smo.sorted.bam</span></tt></li>
</ul>
</div>
<div class="section" id="merging-sam-files">
<h2>Merging sam files<a class="headerlink" href="#merging-sam-files" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">Fl1-1.bam</span></tt> and <tt class="docutils literal"><span class="pre">FL1-2.bam</span></tt> are two technical replicates of the one control sample FL1. We would like to merge these two sam files.</p>
<p>Getting help</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools merge
</pre></div>
</div>
</div></blockquote>
<p>Note the usage and run the command</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools merge -h header.sam ~/FL1-merged.bam rawData/FL1-1.bam rawData/FL1-2.bam
</pre></div>
</div>
</div></blockquote>
<ul class="simple">
<li>What does <tt class="docutils literal"><span class="pre">-h</span></tt> in the above command do?</li>
<li>Does this command work? If not, why not? What command can allow the files to be merged.</li>
</ul>
</div>
<div class="section" id="get-summary-statistics">
<h2>Get summary statistics<a class="headerlink" href="#get-summary-statistics" title="Permalink to this headline">¶</a></h2>
<p>Use <tt class="docutils literal"><span class="pre">samtools</span> <span class="pre">idxstats</span></tt> to get summary statistics for the aligned file. Use the help.</p>
</div>
<div class="section" id="create-a-fasta-file-index">
<h2>Create a fasta file index<a class="headerlink" href="#create-a-fasta-file-index" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools faidx genome/dros_BDGP5.25.fa
</pre></div>
</div>
</div></blockquote>
<ul class="simple">
<li>What is the benefit of creating an index of a fasta file?</li>
<li>Does this command work? Why not?</li>
</ul>
</div>
<div class="section" id="create-a-pileup">
<h2>Create a pileup<a class="headerlink" href="#create-a-pileup" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">samtools</span> <span class="pre">mpileup</span></tt> is a very useful utility for calling variants in alignment files. Read the help documentation carefully.</p>
<blockquote>
<div><div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">This command is slow.</p>
</div>
<div class="highlight-bash"><div class="highlight"><pre>&gt; samtools mpileup -g -l intervalFile.bed -I -D -q 20 -f genome/dros_BDGP5.25.fa rawData/arm.bam
</pre></div>
</div>
</div></blockquote>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><tt class="docutils literal"><span class="pre">dros_BDGP5.25.fa</span></tt> needs to be indexed otherwise the above command will not work</p>
</div>
</div>
<div class="section" id="getting-bored-good-to-know">
<h2>Getting bored? Good to know...<a class="headerlink" href="#getting-bored-good-to-know" title="Permalink to this headline">¶</a></h2>
<p><strong>Working with the stream</strong>
You can take the output of one command from the “standard stream” and pipe (see section <a class="reference internal" href="../introduction.html#pipe"><em>Pipe</em></a>) it as input to another samtools command.</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt; samtools view -u arm.bam chrX | samtools pileup -cf dros_BD5.25.fa -
</pre></div>
</div>
</div></blockquote>
<p>What exactly does the above command do?</p>
<p>Use <tt class="docutils literal"><span class="pre">samtools</span></tt> and <tt class="docutils literal"><span class="pre">awk</span></tt> to count the number of mapped reads in your file.</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>&gt;  samtools idxstats rawData/smo.bam | awk <span class="s1">&#39;{s+=$3} END {print s}&#39;</span>
</pre></div>
</div>
</div></blockquote>
<ul class="simple">
<li>Do you get the same number of mapped reads as with <tt class="docutils literal"><span class="pre">samtools</span> <span class="pre">view</span> <span class="pre">-c</span> <span class="pre">smo.bam</span></tt></li>
<li>What is the <tt class="docutils literal"><span class="pre">awk</span></tt> command doing in the above command?</li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/HUBLogo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">SAMtools</a><ul>
<li><a class="reference internal" href="#view-bam-files">View bam files</a></li>
<li><a class="reference internal" href="#viewing-the-header-of-a-bam-file">Viewing the header of a bam file</a></li>
<li><a class="reference internal" href="#count-the-number-of-alignments-in-a-bam-file">Count the number of alignments in a bam file</a></li>
<li><a class="reference internal" href="#create-a-bam-index">Create a bam index</a></li>
<li><a class="reference internal" href="#sorting-the-sam-file">Sorting the sam file</a></li>
<li><a class="reference internal" href="#merging-sam-files">Merging sam files</a></li>
<li><a class="reference internal" href="#get-summary-statistics">Get summary statistics</a></li>
<li><a class="reference internal" href="#create-a-fasta-file-index">Create a fasta file index</a></li>
<li><a class="reference internal" href="#create-a-pileup">Create a pileup</a></li>
<li><a class="reference internal" href="#getting-bored-good-to-know">Getting bored? Good to know...</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="STEP2_alignRNASeqReads.html"
                        title="previous chapter">Aligning RNASeq Reads</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="STEP4_CUFFLINKS_DE.html"
                        title="next chapter">Estimating Expression</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/exercises/STEP3_viewManipulateAlignmentData_samtools.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="STEP4_CUFFLINKS_DE.html" title="Estimating Expression"
             >next</a> |</li>
        <li class="right" >
          <a href="STEP2_alignRNASeqReads.html" title="Aligning RNASeq Reads"
             >previous</a> |</li>
        <li><a href="../index.html">None</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Grainne Kerr, Holger Dinkel, Jon Fuller, Matt Betts.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>