% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\usepackage[utf8]{inputenc}
\DeclareUnicodeCharacter{00A0}{\nobreakspace}
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}


\title{IntroToCommandLine Documentation}
\date{September 18, 2013}
\release{1}
\author{Grainne Kerr, Holger Dinkel, Matt Betts, Jon Fuller}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\expandafter\def\csname PYG@tok@gd\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gu\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@gt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PYG@tok@gs\endcsname{\let\PYG@bf=\textbf}
\expandafter\def\csname PYG@tok@gr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@cm\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@vg\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@m\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@cs\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\colorbox[rgb]{1.00,0.94,0.94}{\strut ##1}}}
\expandafter\def\csname PYG@tok@ge\endcsname{\let\PYG@it=\textit}
\expandafter\def\csname PYG@tok@vc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@il\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@go\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.20,0.20,0.20}{##1}}}
\expandafter\def\csname PYG@tok@cp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@gi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gh\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@ni\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\expandafter\def\csname PYG@tok@nl\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\expandafter\def\csname PYG@tok@nn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@no\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\expandafter\def\csname PYG@tok@na\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@nd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\expandafter\def\csname PYG@tok@ne\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\expandafter\def\csname PYG@tok@si\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\expandafter\def\csname PYG@tok@s2\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@vi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@nt\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\expandafter\def\csname PYG@tok@nv\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@s1\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@gp\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@sh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ow\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@sx\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@bp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c1\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@kc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@mf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@err\endcsname{\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PYG@tok@kd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@ss\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\expandafter\def\csname PYG@tok@sr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\expandafter\def\csname PYG@tok@mo\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@o\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PYG@tok@kr\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@s\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@kp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@w\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PYG@tok@kt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\expandafter\def\csname PYG@tok@sc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@k\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@se\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sd\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZam{\char`\&}
\def\PYGZlt{\char`\<}
\def\PYGZgt{\char`\>}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZhy{\char`\-}
\def\PYGZsq{\char`\'}
\def\PYGZdq{\char`\"}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{IntroToCommandLine::doc}


Contents:


\chapter{Introduction to the Linux Commandline}
\label{introduction:welcome-to-introduction-to-the-command-line-for-ngs-documentation}\label{introduction:introduction-to-the-linux-commandline}\label{introduction::doc}

\section{Why Use the Commandline}
\label{introduction:why-use-the-commandline}\begin{itemize}
\item {} 
It's \textbf{fast}. Productivity is a word that gets tossed around a lot by so-called power users, but the command line can really streamline your computer use, assuming you learn to use it right.

\item {} 
It's \textbf{easier to get help}. The command line may not be the easiest thing to use, but it makes life a whole lot easier for people trying to help you and for yourself when looking for help, especially over the internet. Many times it's as simple as the helper posting a few commands and some instructions and the recipient copying and pasting those commands. Anyone who has spent hours listening to someone from tech support say something like, “OK, now click this, then this, then select this menu command” knows how frustrating the GUI alternative can be.

\item {} 
It's nearly \textbf{universal}. There are hundreds of Linux distros out there, each with a slightly different graphical environment. Thankfully, the various distros do have one common element: the command line. There are distro-specific commands, but the bulk of commands will work on any Linux system.

\item {} 
It's \textbf{powerful}. The companies behind those other operating systems try their best to stop a user from accidentally screwing up their computer. Doing this involves hiding a lot of the components and tools that could harm a computer away from novices. Linux is more of an open book, which is due in part to its prominent use of the command line.

\end{itemize}


\section{General Remarks Regarding Using UNIX/Linux Systems}
\label{introduction:general-remarks-regarding-using-unix-linux-systems}\begin{itemize}
\item {} 
\textbf{Test before run}. Anything written here has to be taken with a grain of salt. On another system – be it a different Linux distribution or another UNIXoid operating system – you might find the same command but without the support of some of the options taught here. It is even possible, that the same option has a different meaning on another system. With this in mind always make sure to test your commands (specially the “dangerous” ones which remove or modify files) when switching from one system to the other.

\item {} 
\textbf{The Linux/UNIX environment}. The behaviour of many commands is influenced or controlled by the so-called “environment”. This environment is the sum of all your environment variables. Some of these environment variables will be shown towards the end of this course.

\item {} 
\textbf{UPPERCASE, lowercase}. Don’t forget that everything is case-sensitive.

\item {} 
\textbf{The Filesystem}. Linux filesystems start on top at the root directory (sic!) “/” which hierarchically broadens towards the ground.  The separator between directories or directories and files in Linux is the slash (“/”).

\end{itemize}
\begin{figure}[htbp]
\centering
\capstart

\includegraphics{_static/filesystem.png}
\caption{Depending on the Linux distribution you might or might not find all of above
directories. Most important directories for you are \code{/bin} and \code{/usr/bin}
(sometimes also \code{/usr/local/bin}) which contain the user software, \code{/home} which
usually contains the users’ homedirectories and \code{/tmp} which can be used to store
temporary data (beware: Its content is regularly removed!).}\end{figure}

..note:: The terms “directory” and “folder” are used interchangeably in this document.


\section{General Structure of Linux Commands}
\label{introduction:general-structure-of-linux-commands}
Linux commands have the following general structure:
\begin{figure}[htbp]
\centering

\includegraphics{_static/LinuxCommandStructure.png}
\end{figure}

commandline options (sometimes called comandline switches) commonly have one of the two following forms: The short form \code{-character} or the long form \code{-{-}string.} E.g.
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} man -h
\textgreater{} man --help
\end{Verbatim}
\end{quote}

Short options are usually – though not always – concatenable:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ls -l -A -h
\textgreater{} ls -lAh
\end{Verbatim}
\end{quote}

Some options require an additional argument, which is added with a blank to the short form and with an equal sign to the long form:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ls -I “*.pdf”
\textgreater{} ls --ignore=”*.pdf”
\end{Verbatim}
\end{quote}

Since Linux incorporates commands from different sources, options can be available in one or both forms and you’ll also encounter options with no dash at all and all kinds of mixtures:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} tar cf file.tar -C .. file/
\textgreater{} ps auxgww
\end{Verbatim}
\end{quote}


\subsection{A Journey Through the Commands}
\label{introduction:a-journey-through-the-commands}
Please note that all examples and usage instructions below are just a glimpse of what you can do and reflect our opinion on what’s important and what’s not. Most of these commands support many more options and different usages. Consult the manpages to find them.
Typographical conventions: Commands and examples are written in Courier.  User Input is written in Courier bold and placeholders are generally written in italic.


\section{Getting Help}
\label{introduction:getting-help}\label{introduction:help}
\code{-h/-{-}help} option, no parameters

Many commands support a “help” option, either through \code{-h} or through \code{-{-}help}.
Other commands will show a help page or at least a short usage overview if you provide wrong commandline options

\textbf{Usage}:
\code{man command or file}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} man man
man(1)

NAME
 man - format and display the on-line manual pages

SYNOPSIS
 man [-acdfFhkKtwW] [--path] [-m system] [-p string] [-C config\_file]
[...]
\end{Verbatim}
\end{quote}

For the navigation within a \index{man}man-page see the chapter regarding less below.

\begin{notice}{note}{Note:}
The behaviour of man is dependent of the \$PAGER environment variable
\end{notice}

\textbf{Usage}: \code{apropos keyword}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} apropos who
[...]
\textgreater{} who                  (1)  - show who is logged on
\textgreater{} who                  (1p)  - display who is on the system
\textgreater{} whoami               (1)  - print effective userid
\end{Verbatim}
\end{quote}

Use \index{apropos}apropos to find candidates for specific tasks

The \code{/usr/share/doc} directory in some Linux distributions contains additional documentation of installed software packages


\section{Who am I, where am I}
\label{introduction:who-am-i-where-am-i}
\textbf{Usage}: \code{whoami}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} whoami
fthommen
\end{Verbatim}
\end{quote}

\textbf{Usage}: \code{hostname}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} hostname
pc-teach01
\end{Verbatim}
\end{quote}

\textbf{Usage}: \code{pwd}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} pwd
/home/fthommen
\end{Verbatim}
\end{quote}

\textbf{Usage}: \code{date}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} date
Tue Sep 25 19:57:50 CEST 2012
\end{Verbatim}
\end{quote}

\begin{notice}{note}{Note:}
The command time does something completely different than \index{date}date and is not used to show the current time.
\end{notice}

\textbf{Usage}: \code{cd {[}new\_directory{]}}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\# pwd
/home/fthommen
\# cd /usr/bin
\# pwd
/usr/bin
\end{Verbatim}
\end{quote}

Special directories:
\begin{itemize}
\item {} 
“\code{.}”:  The current working directory

\item {} 
“\code{..}”: The parent directory of the current working directory

\item {} 
“\code{\textasciitilde{}}”:  Your homedirectory

\end{itemize}

\begin{notice}{note}{Note:}
Using \index{cd}cd without a directory is equivalent to “\code{cd \textasciitilde{}}” and changes into the users’s homedirectory
\end{notice}

\begin{notice}{note}{Note:}
Please note the difference between absolute pathes (starting with ``\code{/}'') and relative pathes (starting with a directory name)
\end{notice}

\begin{Verbatim}[commandchars=\\\{\}]
\$ pwd
/usr
\$ cd /bin
\$ pwd
/bin
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} pwd
/usr
\textgreater{} cd bin
\textgreater{} pwd
/usr/bin
\end{Verbatim}
\begin{description}
\item[{\textbf{Usage}:}] \leavevmode
\code{ls {[}options{]} {[}file(s) or directory/ies{]}}

\end{description}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ls
/home/fthommen
\textgreater{} ls –l aa.pdf
-rw-r--r-- 1 fthommen cmueller 0 Sep 24 10:59 aa.pdf
\end{Verbatim}

Useful options:
\begin{optionlist}{3cm}
\item [-l]  
Long listing with permissions, user, group and last modification date
\item [-1]  
Print listing in one column only
\item [-a]  
Show all files (hidden, “.” and “..”)
\item [-A]  
Show almost all files (hidden, but not “.” and “..”)
\item [-F]  
Show filetypes (nothing = regular file, “/” = directory, “*” = executable file, “@” = symbolic link)
\item [-d]  
Show directory information instead of  directory content
\item [-t]  
Sort listing by modification time (most recent on top)
\end{optionlist}


\bigskip\hrule{}\bigskip

\begin{figure}[htbp]
\centering

\includegraphics{_static/LongListingDeconstructed.png}
\end{figure}


\bigskip\hrule{}\bigskip


Files and folders can’t only be referred to with their full name, but also with so-called “Shell Globs”, which are a kind of simple pattern to address groups of files and folders.  Instead of explicit names you can use the following placeholders:
\begin{itemize}
\item {} 
\code{?:}  Any single character

\item {} 
\code{*:}  Any number of any character (including no character at all)

\item {} 
\code{{[}...{]}:}    One of the characters included in the brackets.  Use “-“ to define ranges  of characters

\end{itemize}

Examples:
\begin{itemize}
\item {} 
\code{*.pdf:}  All files having the extension “.pdf”

\item {} 
\code{?.jpg:}  Jpeg file consisting of only one character

\item {} 
\code{{[}0-9{]}*.txt:} All files starting with a number and having the extension “.txt”

\item {} 
\code{*.???:}  All files having a three-character extension

\end{itemize}

\begin{notice}{note}{Note:}
The special directory “\code{\textasciitilde{}}” mentioned above is a shell glob, too.
\end{notice}

\textbf{Usage}:  \code{touch file(s) or directory/ies}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ls afile
ls: afile: No such file or directory
\textgreater{} touch afile
\textgreater{} ls afile
afile
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ls –l aa.pdf
-rw-r--r-- 1 fthommen cmueller 0 Sep 24 10:59 aa.pdf
\textgreater{} touch aa.pdf
\textgreater{} ls –l aa.pdf
-rw-r--r-- 1 fthommen cmueller 0 Sep 25 22:01 aa.pdf
\end{Verbatim}
\end{quote}

\textbf{Usage}:  \code{rm {[}options{]} file(s)}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
rm –r [options] directory/ies
\textgreater{} ls afile
afile
\textgreater{} rm afile
\textgreater{} ls afile
ls: afile: No such file or directory
\end{Verbatim}
\end{quote}

\textbf{Useful options}:
\begin{optionlist}{3cm}
\item [-i]  
Ask for confirmation of each removal
\item [-r]  
Remove recursively
\item [-f]  
Force the removal (no questions, no errors if a file doesn't exist)
\end{optionlist}

\begin{notice}{note}{Note:}
rm without the –i option will usually not ask you if you really want to remove the file or directory
\end{notice}

\textbf{Usage}:  \code{mv {[}options{]} sourcefile destinationfile}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
mv [options] sourcefile(s) destinationdirectory
\textgreater{} ls *.txt
a.txt
\textgreater{} mv a.txt b.txt
\textgreater{} ls *.txt
b.txt
\end{Verbatim}
\end{quote}

\textbf{Useful options}:
\begin{optionlist}{3cm}
\item [-i]  
Ask for confirmation of each removal
\end{optionlist}

\begin{notice}{note}{Note:}
You cannot overwrite an existing directory by another one with mv
\end{notice}

\textbf{Usage}:  \code{mkdir {[}options{]} directory}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ls adir/
ls: adir/: No such file or directory
\textgreater{} mkdir adir
\textgreater{} ls adir
\end{Verbatim}
\end{quote}

\textbf{Useful options}:
\begin{optionlist}{3cm}
\item [-p]  
Create parent directories (when creating nested directories)

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} mkdir adir/bdir
mkdir: cannot create directory {}`adir/bdir': No such file or directory
\textgreater{} mkdir –p adir/bdir
\end{Verbatim}
\end{optionlist}

\textbf{Usage}:  \code{rmdir directory}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} rmdir adir/
\end{Verbatim}
\end{quote}

\begin{notice}{note}{Note:}
If the directory is not empty, rmdir will complain and not remove it
\end{notice}

\textbf{Usage}:  \code{cp {[}options{]} sourcefile destinationfile}
.. note:: \index{cp}cp {[}options{]} sourcefile(s) destinationdirectory
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} cp P12931.fasta backup\_of\_P12931.fasta
\end{Verbatim}
\end{quote}

\textbf{Useful options}:
\begin{optionlist}{3cm}
\item [-r]  
Copy recursively
\item [-i]  
Interactive operation, ask before overwriting an existing file
\item [-p]  
Preserve owner, permissions and timestamp
\end{optionlist}

\textbf{Usage}:  \code{cat {[}options{]} file(s)}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} cat  P12931.fasta backup\_of\_P12931.fasta
[...]
\end{Verbatim}
\end{quote}

\begin{notice}{note}{Note:}
\index{cat}cat only makes sense for short files or for e.g. combining several files into one.  See the redirection examples later
\end{notice}

\textbf{Usage}:  \code{less {[}options{]} file(s)}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} less  P12931.fasta backup\_of\_P12931.fasta
[...]
\end{Verbatim}
\end{quote}

\begin{notice}{note}{Note:}
This is the default “pager” for manpages under Linux unless you redefine your \$PAGER environment variable
\end{notice}

\textbf{Navigation within less}:

\begin{tabulary}{\linewidth}{|L|L|}
\hline
\textsf{\relax 
Key(s):
} & \textsf{\relax 
Effect:
}\\
\hline
up, down, right, left:
 & 
use cursor keys
\\

top of document:
 & 
g
\\

bottom of document:
 & 
G
\\

search:
 & 
“/” + search-term
\\

find next match:
 & 
n
\\

find previous match:
 & 
N
\\

quit:
 & 
q
\\
\hline\end{tabulary}


\index{grep}Grep is a command-line utility for searching plain-text data sets for lines matching a regular expression.

\textbf{Usage}:  \code{grep {[}options{]} pattern file(s)}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} grep -i ensembl P04637.txt
DR   Ensembl; ENST00000269305; ENSP00000269305; ENSG00000141510.
DR   Ensembl; ENST00000359597; ENSP00000352610; ENSG00000141510.
DR   Ensembl; ENST00000419024; ENSP00000402130; ENSG00000141510.
DR   Ensembl; ENST00000420246; ENSP00000391127; ENSG00000141510.
DR   Ensembl; ENST00000445888; ENSP00000391478; ENSG00000141510.
DR   Ensembl; ENST00000455263; ENSP00000398846; ENSG00000141510.
\end{Verbatim}
\end{quote}

\textbf{Useful options}:
\begin{optionlist}{3cm}
\item [-v]  
Print lines that do not match
\item [-i]  
Search case-insensitive
\item [-l]  
List files with matching lines, not the lines itself
\item [-L]  
List files without matches
\item [-c]  
Print count of matching lines for each file
\end{optionlist}

\index{head}Head is a program on Unix and Unix-like systems used to display the beginning of a text file or piped data.

\textbf{Usage}:  \code{head {[}options{]} file(s)}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} head /etc/passwd
root:x:0:0:root:/root:/bin/bash
bin:x:1:1:bin:/bin:/sbin/nologin
daemon:x:2:2:daemon:/sbin:/sbin/nologin
adm:x:3:4:adm:/var/adm:/sbin/nologin
lp:x:4:7:lp:/var/spool/lpd:/sbin/nologin
sync:x:5:0:sync:/sbin:/bin/sync
shutdown:x:6:0:shutdown:/sbin:/sbin/shutdown
halt:x:7:0:halt:/sbin:/sbin/halt
mail:x:8:12:mail:/var/spool/mail:/sbin/nologin
news:x:9:13:news:/etc/news:
\end{Verbatim}
\end{quote}

\textbf{Useful options}:
\begin{optionlist}{3cm}
\item [-n num]  
Print num lines (default is 10)
\end{optionlist}

\textbf{Usage}:  \code{tail {[}options{]} file(s)}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} tail –n 3 /etc/passwd
xfs:x:43:43:X Font Server:/etc/X11/fs:/sbin/nologin
gdm:x:42:42::/var/gdm:/sbin/nologin
sabayon:x:86:86:Sabayon user:/home/sabayon:/sbin/nologin
\end{Verbatim}
\end{quote}

\textbf{Useful options}:
\begin{optionlist}{3cm}
\item [-n num]  
Print num lines (default is 10)
\item [-f]  
``Follow'' a file (print new lines as they are written to the file)
\end{optionlist}

\textbf{Usage}:  \code{file {[}options{]} file(s)}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} file /bin/date
/bin/date: ELF 32-bit LSB executable
\textgreater{} file /bin
/bin: directory
\textgreater{} file SRC\_HUMAN.fasta
SRC\_HUMAN.fasta: ASCII text
\end{Verbatim}
\end{quote}

\begin{notice}{note}{Note:}
The command \index{file}file uses certain tests and some magic to determine the type of a file
\end{notice}

\textbf{Usage}:  \code{which {[}options{]} command(s)}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} which date
/bin/date
\textgreater{} which eclipse
/usr/bin/eclipse
\textgreater{}
\end{Verbatim}
\end{quote}

\textbf{Usage}:  \code{find {[}starting path(es){]} {[}search filter{]}}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} find /etc
/etc
/etc/printcap
/etc/protocols
/etc/xinetd.d
/etc/xinetd.d/ktalk
[...]
\textgreater{}
\end{Verbatim}
\end{quote}

\code{find} \index{find}is a powerful command with lots of possible search filters.  Refer to the manpage for a complete list.

\textbf{Examples}:
\begin{itemize}
\item {} 
Find by name:

\end{itemize}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} find . -name SRC\_HUMAN.fasta
./SRC\_HUMAN.fasta
\end{Verbatim}
\end{quote}
\begin{itemize}
\item {} 
Find by size: (List those entries in the directory /usr/bin that are bigger than 500kBytes)

\end{itemize}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} find /usr/bin -size +500k
/usr/bin/oparchive
/usr/bin/kiconedit
/usr/bin/opjitconv
[...]
\end{Verbatim}
\end{quote}
\begin{itemize}
\item {} 
Find by type (d=directory, f=file, l=link)

\end{itemize}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} find . -type d
.
./adir
\end{Verbatim}
\end{quote}

\textbf{Usage}:  \code{clear}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} clear
\end{Verbatim}
\end{quote}

In case the output of the terminal/screen gets cluttered, you can use \code{clear} to \index{clear}clear the screen...

If this doesn't work, you can use \code{reset} to perform a re-initialization of the terminal:

\textbf{Usage}:  \code{reset {[}options{]}}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} reset
\end{Verbatim}
\end{quote}

using ls -l to view entries of current directory:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ls –l
drwxr-xr-x 2 dinkel gibson 4096 Sep 17 10:46 adir
lrwxrwxrwx 1 dinkel gibson   15 Sep 17 10:45 H1.fasta -\textgreater{} H2.fasta
-rw-r--r-- 1 dinkel gibson  643 Sep 17 10:45 H2.fasta
\end{Verbatim}
\end{quote}
\begin{figure}[htbp]
\centering

\includegraphics{_static/LinuxPermissions.png}
\end{figure}

Permissions are set using the \code{chmod} (\index{chmod}change mode ) command.
\textbf{Usage}:  \code{chmod {[}options{]} mode(s) files(s)}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ls –l adir
drwxr-xr-x 2 dinkel gibson 4096 Sep 17 10:46 adir
\textgreater{} chmod u-w,o=w adir
\textgreater{} ls –l adir
dr-xr-x-w- 2 dinkel gibson 4096 Sep 17 10:46 adir
\end{Verbatim}
\end{quote}

The mode is composed of

\begin{tabulary}{\linewidth}{|L|L|L|L|L|L|L|}
\hline
 \multicolumn{2}{|l|}{\textsf{\relax 
Who
}} &  \multicolumn{3}{l|}{\textsf{\relax 
What
}} &  \multicolumn{2}{l|}{\textsf{\relax 
Which permission
}}\\
\hline
u:
 & 
user/owner
 &  \multicolumn{2}{l|}{
+:
} & 
add this permission
 & 
r:
 & 
read
\\

g:
 & 
group
 &  \multicolumn{2}{l|}{
-:
} & 
remove this permission
 & 
w:
 & 
write
\\

o:
 & 
other
 &  \multicolumn{2}{l|}{
=:
} & 
set exactly this permission
 & 
x:
 & 
execute
\\

a:
 &  \multicolumn{3}{l|}{
all          \textbar{}  xx
} & 
xx
 & 
xx
 & 
xx
\\
\hline\end{tabulary}


Add executable permission to the group:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} chmod g+x file
\end{Verbatim}
\end{quote}

Revoke this permission:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} chmod g-x file
\end{Verbatim}
\end{quote}

Allow all to read a directory:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} chmod a+rx adir/
\end{Verbatim}
\end{quote}

To execute commands at a remote machine/server, you need to log in to this machine. This is done using the \code{ssh} command (secure shell).
In its simplest form, it takes just the machinename as parameter (assuming the username on the local machine and remote machine are identical):
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ssh remote\_server
\end{Verbatim}
\end{quote}

\begin{notice}{note}{Note:}
Once logged in, \index{hostname}\index{whoami}use \code{hostname,} \code{whoami,} etc. to determine on which machine you are currently working!
\end{notice}

To use a different username, you can use either:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ssh username@remote\_server
\end{Verbatim}
\end{quote}

or
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ssh -l username remote\_server
\end{Verbatim}
\end{quote}

When connecting to a machine for the first time, it might display a warning:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} ssh sub-master
The authenticity of host 'sub-master (10.11.4.84)' can't be established.
RSA key fingerprint is 47:a4:0f:7b:c2:0f:ef:91:8e:65:fc:3c:f7:0c:53:8d.
Are you sure you want to continue connecting (yes/no)?
\end{Verbatim}
\end{quote}

Type \emph{yes} here.
If this message appears a second time, you should contact your IT specialist...
\begin{description}
\item[{To \index{exit}\index{disconnect}disconnect from the remote machine, type:}] \leavevmode
\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} exit
\end{Verbatim}

\end{description}

Copying files to and from remote computers can be done using \code{scp}  (secure copy).
The order of parameters is the same as in \code{cp}: first the name of the source, then the name of the destination. Either one can be the remote part.
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} scp localfile server:/remotefile

\textgreater{} scp server:/remotefile localfile
\end{Verbatim}
\end{quote}

An alternative username can be provided just as in ssh:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} scp username@server:/remotefile localfile
\end{Verbatim}
\end{quote}

\index{redirect}Redirect the output of one program into e.g. a file: (Caution: you can easily overwrite files by this!)
Inserting the current date into a new file:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} date \textgreater{} file\_containing\_date
\end{Verbatim}
\end{quote}

\index{grep}Filtering lines containing the term “src” from FASTA files and inserting them into the file lines\_with\_src.txt:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} cd /exercises/
\textgreater{} grep -i "src" *.fasta \textgreater{} lines\_with\_src.txt
\end{Verbatim}
\end{quote}

\index{append}Append something to a file (rather than overwriting it):
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} date \textgreater{}\textgreater{} file\_containing\_date
\end{Verbatim}
\end{quote}

Use the \index{\textbar{}}\index{pipe}\textbar{} pipe symbol (\code{\textbar{}}) to feed the output of one program into the next program.
Here: use \code{ls} to show the directory contents and then \index{grep}use \code{grep} to only show those that contain fasta in their name:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} cd /exercises
\textgreater{} ls \textbar{} grep fasta
EPSINS.fasta
FYN\_HUMAN.fasta
P12931.fasta
SRC\_HUMAN.fasta
\end{Verbatim}
\end{quote}

\index{environment variables}Environment variables are a set of dynamic named values that can affect the way running processes will behave on a computer.

Contains the location of the user's home directory. Although the current user's home directory can also be found out through the
C functions \code{getpwuid} and \code{getuid,} \index{\$HOME}\$HOME is often used for convenience in various shell scripts (and other contexts).

\begin{notice}{note}{Note:}
Do not change this variable unless you have a good reason and you know what you are doing!
\end{notice}

\index{\$PATH}\$PATH contains a colon-separated list of directories that the shell searches for commands that do not contain a slash in their name
(commands with slashes are interpreted as file names to execute, and the shell attempts to execute the files directly).

The \index{\$PAGER}\$PAGER variable contains the path to the program used to list the contents of files through (such \index{less}as \code{less} \index{more}or \code{more}).

The \index{\$PWD}\$PWD variable points to the current directory. Equivalent to the output of the command \code{pwd} when called without arguments.

\index{environment variables!display}Use \code{echo} to \index{echo}display individual variables \emph{set{}`} \index{env}or \code{env} to view all at once:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} echo \$HOME
/localhome/teach01
\textgreater{} set
...
\textgreater{} env
...
\end{Verbatim}
\end{quote}

Use \code{export} followed by the variable name and the value of the variable (separated by the equal sign) to \index{environment variables!set}\index{set}\index{export}set an environment variable:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} export PAGER=/usr/bin/less
\end{Verbatim}
\end{quote}

\begin{notice}{note}{Note:}
An environment variable is only valid for your current session. Once you logout of your current session, it is lost or reset.
\end{notice}


\chapter{Exercises}
\label{exercises_beginner:exercises}\label{exercises_beginner::doc}

\section{Misc. file tools}
\label{exercises_beginner:misc-file-tools}\begin{enumerate}
\item {} 
Which tool can be used to determine the type of a file?

\item {} 
Use it on the following files/directories and compare the results:
\begin{enumerate}
\item {} 
\code{/usr/bin/tail}

\item {} 
\code{\textasciitilde{}}

\item {} 
\code{/exercises/SRC\_HUMAN.fasta}

\end{enumerate}

\end{enumerate}


\section{Searching}
\label{exercises_beginner:searching}\begin{enumerate}
\item {} 
Which tool can be used to search for files or directories?

\item {} 
Use it to find all directories in the \code{/exercises} directory

\item {} 
Search for the file date in the \code{/bin} directory

\item {} 
List those entries in the directory \code{/bin} that are bigger than 400kBytes

\end{enumerate}


\section{Misc. terminal}
\label{exercises_beginner:misc-terminal}\begin{enumerate}
\item {} 
Which two tools can be used to redraw/empty the screen?

\end{enumerate}


\section{Permissions}
\label{exercises_beginner:permissions}\begin{enumerate}
\item {} 
Create a directory called testpermissions

\item {} 
Change your working directory to testpermissions

\item {} 
Create a directory called adir.

\item {} 
Use the command \code{which date} to find out where the date program is located.

\item {} 
Copy this date program into the directory adir.

\item {} 
Check the permissions of the copied program date

\item {} 
Change the permissions on date to remove the executable permissions.

\item {} 
Check the permissions of the program date

\item {} 
Try running it as \code{./date} or \code{adir/date} (depending on your current working directory)

\item {} 
Change the permissions back so that the file is executable.

\item {} 
Try running it as \code{./date} or \code{adir/date} (depending on your current working directory)

\item {} 
Copy a textfile from a previous exercise into adir, then change the permissions, so you are not allowed to write to it.

\item {} 
Then change the permissions so you can't read/cat it either.

\item {} 
Change your working directory to testpermissions, and then try changing the permissions on adir.

\item {} 
What are the minimum permissions (on the directory) necessary for you to be able to execute \code{adir/date}?

\end{enumerate}


\section{Remote access}
\label{exercises_beginner:remote-access}\begin{enumerate}
\item {} 
Login to machine ``sub-master.embl.de'' (using your own username)

\item {} 
Use exit to quit the remote shell (Beware to not exit your local shell)

\item {} 
Use clear to empty the screen after logout from the remote server

\item {} 
Use the following commands locally as well as on the remote machine to get a feeling for the different machines:

\end{enumerate}
\begin{enumerate}
\item {} 
\code{hostname}

\item {} 
\code{whoami}

\item {} 
\code{cat /etc/hostname}

\item {} 
\code{ls -la \textasciitilde{}/}

\end{enumerate}
\begin{enumerate}
\item {} 
Copy the file \code{/etc/motd} from machine sub-master.embl.de into your local home directory

\item {} 
Determine the filetype and the permissions of the file that you just copied

\item {} 
Login to your neighbor's machine (ask them for the hostname) using the username \code{teach01} (password will be given by teacher)

\end{enumerate}


\section{IO and Redirections}
\label{exercises_beginner:io-and-redirections}\begin{enumerate}
\item {} 
Use date in conjuction with the redirection to insert the current date into the (new) file current\_date (in your homedirectory).

\item {} 
Inspect the file to make sure it contains (only a single line with) the date.

\item {} 
Use \code{date} again to append the current date into the same file.

\item {} 
Again, check that this file now contains two lines with dates.

\item {} 
Use \code{grep} to filter out lines containing the term ``TITLE'' from all PDB files in the exercises directory and use redirection to insert them into a new file pdb\_titles.txt.

\item {} 
(OPTIONAL) Upon inspection of the file pdb\_titles.txt, you see that it also contains the names of the files in which the term was found.
Use either the grep manpage or \code{grep -{-}help} to find out how you can suppress this behaviour.
Redo the previous exercise such that the output file pdb\_titles.txt only contains lines starting with TITLE.

\end{enumerate}


\chapter{File Formats}
\label{known_file_formats:file-formats}\label{known_file_formats::doc}
File Types to research and present:
\begin{itemize}
\item {} 
Fasta/fastq (sanger and illumina)/fai

\item {} 
GFF/GTF

\item {} 
Sam/Bam/Bai

\item {} 
Vcf/bcf/Pileup/interval/ROD

\item {} 
Bed/bigBed

\item {} 
Wig/bigWig

\end{itemize}


\section{Presentation Task}
\label{known_file_formats:presentation-task}
Give a white-board/flip-chart presentation on one of the above file types. Your presentation should be 5 - 7 mins.
Questions you could answer in your presentation:
\begin{itemize}
\item {} 
\textbf{What does each line in the file represent?}

\item {} 
\textbf{What does each column represent?}

\item {} 
\textbf{What character separates the columns? Can you write this character?}

\item {} 
\textbf{Is there a file header?}

\item {} 
\textbf{What online repositories are available to download this data and how can I download it?}

\item {} 
What type of user/database/online tools uses this file format/who came up with it?

\item {} 
Show an example of the file.

\item {} 
What is the difference between the files you’ve been given?

\item {} 
Can you convert between one file format and another?

\item {} 
Any other information interesting information?

\end{itemize}

\begin{notice}{note}{Note:}
Questions in bold should be asked about any file format you have been given.
\end{notice}

Unsure about jargon? Ask for help!


\subsection{Useful links about file formats}
\label{known_file_formats:useful-links-about-file-formats}

\subsubsection{VCF:}
\label{known_file_formats:vcf}
\href{http://www.1000genomes.org/wiki/analysis/variant-call-format/vcf-variant-call-format-version-42}{http://www.1000genomes.org/wiki/analysis/variant-call-format/vcf-variant-call-format-version-42}


\subsubsection{GFF:}
\label{known_file_formats:gff}\begin{itemize}
\item {} 
\href{http://www.ensembl.org/info/website/upload/gff.html}{http://www.ensembl.org/info/website/upload/gff.html}

\item {} 
\href{http://www.sanger.ac.uk/resources/software/gff/spec.html}{http://www.sanger.ac.uk/resources/software/gff/spec.html}

\end{itemize}


\subsubsection{BED/BIGBED/WIG/BIGWIG:}
\label{known_file_formats:bed-bigbed-wig-bigwig}\begin{itemize}
\item {} 
\href{http://genome.ucsc.edu/FAQ/FAQformat.html\#format1}{http://genome.ucsc.edu/FAQ/FAQformat.html\#format1}

\item {} 
\href{http://www.ensembl.org/info/website/upload/bed.html}{http://www.ensembl.org/info/website/upload/bed.html}

\item {} 
\href{http://www.ensembl.org/info/website/upload/wig.html}{http://www.ensembl.org/info/website/upload/wig.html}

\end{itemize}


\subsubsection{SAM/BAM/BAI:}
\label{known_file_formats:sam-bam-bai}\begin{itemize}
\item {} 
\href{http://genome.sph.umich.edu/wiki/SAM}{http://genome.sph.umich.edu/wiki/SAM}

\item {} 
\href{http://samtools.sourceforge.net/SAMv1.pdf}{http://samtools.sourceforge.net/SAMv1.pdf}

\item {} 
\href{http://blog.nextgenetics.net/?e=18}{http://blog.nextgenetics.net/?e=18} = bitwise flags

\item {} 
\href{http://picard.sourceforge.net/explain-flags.html}{http://picard.sourceforge.net/explain-flags.html}

\end{itemize}


\subsubsection{FASTA/FASTQ:}
\label{known_file_formats:fasta-fastq}\begin{itemize}
\item {} 
\href{http://en.wikipedia.org/wiki/FASTQ\_format\#Illumina\_sequence\_identifiers}{http://en.wikipedia.org/wiki/FASTQ\_format\#Illumina\_sequence\_identifiers}

\item {} 
\href{http://bioinf.comav.upv.es/courses/sequence\_analysis/sequence\_file\_formats.html}{http://bioinf.comav.upv.es/courses/sequence\_analysis/sequence\_file\_formats.html}

\item {} 
\href{http://blog.nextgenetics.net/?e=33}{http://blog.nextgenetics.net/?e=33}  = phred quality

\end{itemize}


\subsubsection{All files!}
\label{known_file_formats:all-files}
\href{http://www.broadinstitute.org/igv/FileFormats}{http://www.broadinstitute.org/igv/FileFormats}
\href{http://genome.ucsc.edu/FAQ/FAQformat}{http://genome.ucsc.edu/FAQ/FAQformat}


\subsubsection{NEXT STEPS}
\label{known_file_formats:next-steps}
Look at the databases mentioned in the presentations and go through them in a live demo. Typical
will be Ensembl, UCSC, NCBI, GATK, 1000 genomes

SAM very important – re-enforce QUAL score, CIGAR string, FLAG
Quality scores – make reference to illumina vs sanger (Sanger pretty standard now)


\chapter{Files}
\label{file:files}\label{file::doc}
Raw data:
\begin{itemize}
\item {} 
data/smo.bam

\item {} 
data/arm.bam

\item {} 
data/fl1-1.bam

\item {} 
data/f11-2.bam

\item {} 
data/smo.2Lsubset.fastq

\item {} 
data/arm.Xsubset.fastq

\item {} 
data/dnaSeq1.bam

\item {} 
data/dnaSeq2.bam

\end{itemize}

Gene Reference:
\begin{itemize}
\item {} 
geneRef/dros\_BD5.25.gtf

\item {} 
geneRef/drosophilaMelanogaster.X.gtf

\item {} 
geneRef/drosophilaMelanogaster.2L.gtf

\end{itemize}

Fasta Files:
\begin{itemize}
\item {} 
genome/dros\_BD5.25.fa

\end{itemize}

Bowtie Index Files:
\begin{itemize}
\item {} 
bowtieIndex/drosophilaMelanogaster.2L

\item {} 
bowtieIndex/drosophilaMelanogaster.X

\end{itemize}

Tophat output Files:

Cufflinks Output Files:

Other:
\begin{itemize}
\item {} 
header.sam

\item {} 
intervalFile.bed

\end{itemize}


\chapter{Quality Control}
\label{STEP1_QualityControl:quality-control}\label{STEP1_QualityControl::doc}

\section{In this section you will need to use the following three files:}
\label{STEP1_QualityControl:in-this-section-you-will-need-to-use-the-following-three-files}\begin{itemize}
\item {} 
data/arm.fastq

\item {} 
data/arm.bam

\item {} 
data/FL1-1.sam

\end{itemize}

As a reminder, you can find out what each of these files is used for here.

One of the most important steps before spending lots of time on an analysis, is to check that the raw data from the sequencing run is of good enough quality. We will use a tool called fastqc to do some preliminary quality control of our raw data. The fastqc tool generates summary statistics of sequence and quality data and can be used to filter, reformat and trim next-generation sequence data.

Make a directory in your home directory called \code{fastqcTest}.

Recall, that when we don't know what a tool does, we can always get help (see section {\hyperref[introduction:help]{\emph{Getting Help}}}):

We will use a tool called \textbf{fastqc} to do some preliminary quality control of our raw data. This is a tool that generates summary statistics of sequence and quality data and can be used to filter, reformat and trim next-generation sequence data.

Use your linux know how to make a directory in your home directory called fastqcTest.

Get help on how to use fastqc
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} fastqc -h
\end{Verbatim}
\end{quote}

Report graphs allow us to gain a visual overview of the sequence data. To generate report graphs type the following:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} fastqc -verbose –o fastqcTest –noextract –f fastq arm.fastq -
\end{Verbatim}
\end{quote}
\setbox0\vbox{
\begin{minipage}{0.95\linewidth}
\textbf{Review Questions}

\medskip

\begin{itemize}
\item {} 
What does the option \code{-o} do in the above command?

\item {} 
Does the \code{-o} option have a value associated with it?

\item {} 
What does the -noextract option do? What is the difference between the -o option and the -noextract option?

\end{itemize}
\end{minipage}}
\begin{center}\setlength{\fboxsep}{5pt}\shadowbox{\box0}\end{center}


\subsection{Task}
\label{STEP1_QualityControl:task}
Use fastqc to generate a report of smo.fastq


\section{SAMstat}
\label{STEP1_QualityControl:samstat}
SAMStat is an efficient C program to quickly display statistics of large sequence files from next generation sequencing projects. When applied to SAM/BAM files all statistics are reported for unmapped, poorly and accurately mapped reads separately. This allows for identification of a variety of problems, such as remaining linker and adaptor sequences, causing poor mapping. Apart from this SAMStat can be used to verify individual processing steps in large analysis pipelines.

Use samstat to get statistics on the mapped data file in \emph{arm.fastq} and \emph{arm.bam}

\textbf{Getting help}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samstat -h
\end{Verbatim}
\end{quote}

Note the usage line you are given after typing this
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
Usage:   samstat \textless{}file.sam\textgreater{} \textless{}file.bam\textgreater{} \textless{}file.fa\textgreater{} \textless{}file.fq\textgreater{} ....
\end{Verbatim}
\end{quote}

To get the statistics of the arm.bam file using samstat. This will create a html report in the current working directory
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samstat arm.bam
\end{Verbatim}
\end{quote}


\subsection{Task}
\label{STEP1_QualityControl:id1}
Use samstat to create html report for \code{arm.fastq} and \code{FL1-1.sam}


\chapter{Aligning RNASeq Reads}
\label{STEP2_alignRNASeqReads::doc}\label{STEP2_alignRNASeqReads:aligning-rnaseq-reads}

\section{Example 1 - arm.Xsubset.fastq}
\label{STEP2_alignRNASeqReads:example-1-arm-xsubset-fastq}
This dataset contains reads from a mRNA sample after RNAi knockdown of the arm gene, in the drosophila Melanogaster cell line, S2. The arm gene is found on chromosome X of the drosophila Melanogaster genome. Here, to save time, we will restrict our analysis to this chromosome. You should for your own projects, you use all the genome information.

Your task is to align the reads to the X chromosome

To get information on tophat, open a terminal window and type
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} tophat -h
\end{Verbatim}
\end{quote}

The raw data is called arm.Xsubset.fastq in your home directory, in folder called course\_data \code{arm.Xsubset.fastq}

The annotation Data is for chromosome X is \code{drosophilaMelanogaster.X.gtf}

The bowtie index is \code{bowtieIndex/drosophilaMelanogaster.X}

Note: remember to specify the output directory in the tophat command
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} mkdir -p \textasciitilde{}/course\_data\_output/arm
\end{Verbatim}
\end{quote}

\begin{notice}{note}{Note:}
Specify option --no-coverage-search in the tophat command. This will speed things up.
\end{notice}
\setbox0\vbox{
\begin{minipage}{0.95\linewidth}
\textbf{Review questions}

\medskip

\begin{itemize}
\item {} 
What does tophat do?

\item {} 
What kind of aligner is tophat?

\item {} 
What is the ``bowtie-index''?

\item {} 
How can specifying more mismatches with the -n option change the output?

\item {} 
Why supply a gtf file to tophat?

\item {} 
What are the output files from the tophat aligner?

\item {} 
Why assemble transcripts with cufflinks?

\item {} 
What is the gtf file and why do I need it?

\item {} 
what data does the gtf file contain?

\item {} 
what does the last column in the gtf file contain?

\end{itemize}
\end{minipage}}
\begin{center}\setlength{\fboxsep}{5pt}\shadowbox{\box0}\end{center}


\section{Example 2 - smo.2Lsubset.fastq}
\label{STEP2_alignRNASeqReads:example-2-smo-2lsubset-fastq}
This dataset contains reads from a mRNA sample after RNAi knockdown of the smo gene, in the drosophila Melanogaster cell line, S2.
The arm gene is found on chromosome 2L of the drosophila Melanogaster genome.
Here, to save time, we will restrict our analysis to this chromosome, 2L. You should for your own projects, you use all the genome information.

Your task is to align the reads to the 2L chromosome

To get information on tophat, terminal window and type:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} tophat -h
\end{Verbatim}
\end{quote}

The raw data is called smo.2Lsubset.fastq  \code{data/smo.2Lsubset.fastq}

The annotation Data is for chromosome 2L is: \code{geneRef/drosophilaMelanogaster.2L.gtf}

The bowtie index is \code{bowtieIndex/drosophilaMelanogaster.2L}

\begin{notice}{note}{Note:}
Specify option --no-coverage-search in the tophat command. This will speed things up.
\end{notice}
\setbox0\vbox{
\begin{minipage}{0.95\linewidth}
\textbf{Review questions}

\medskip

\begin{itemize}
\item {} 
What does the -p option in tophat do?

\item {} 
How does --no-coverage search save time?

\item {} 
What does the -g option do?

\end{itemize}
\end{minipage}}
\begin{center}\setlength{\fboxsep}{5pt}\shadowbox{\box0}\end{center}


\chapter{Samtools}
\label{STEP3_viewManipulateAlignmentData_samtools:samtools}\label{STEP3_viewManipulateAlignmentData_samtools::doc}
Website:
\begin{itemize}
\item {} 
\href{http://samtools.sourceforge.net/}{http://samtools.sourceforge.net/}

\item {} 
\href{http://samtools.sourceforge.net/samtools-c.shtml}{http://samtools.sourceforge.net/samtools-c.shtml}

\item {} 
\href{http://samtools.sourceforge.net/samtools.shtml}{http://samtools.sourceforge.net/samtools.shtml}

\end{itemize}

\emph{samtools} is a set of scripts (a toolbox so to say) that can be used to manipulate and view sam/bam files. In particular you can: sort, index, merge, view these files.

\textbf{Getting help}

To print a list of all the tools available in the samtools suite, simply type samtools on the command line
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools
\end{Verbatim}
\end{quote}

To print a list of the parameters required and options available with each tool in the suite, simply type samtools followed by the name of the tool on the command line. For example, to get a list of the options available for the “view” tool in the samtools suite, simply type:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools view
\end{Verbatim}
\end{quote}

\textbf{View bam files:}

smo.bam is a bam file and is not human readable. To make it human readable you can convert it to a sam file.
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools view –h –o smo.sam data/smo.bam
\end{Verbatim}
\end{quote}
\begin{itemize}
\item {} 
What does –h and –o do in the above example?

\item {} 
Convert arm.bam into a sam file.

\end{itemize}

\textbf{Viewing the header of a bam file}

In some cases you might only want to see or generate the header of a bam file.
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools view –H smo.bam
\end{Verbatim}
\end{quote}

Now try:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools view –H smo.bam \textgreater{} smo.header.sam
\end{Verbatim}
\end{quote}
\begin{itemize}
\item {} 
What does the “\textgreater{} smo.header.sam” of the above statement do?

\item {} 
What information is stored in the header of the sam file?

\item {} 
From the header of the file, can you tell which alignment program was used to generate the bam file

\item {} 
Use the samtools view (and read the help) to view a specific region e.g. all reads mapping to chromosome X

\end{itemize}

\textbf{Count the number of alignments in a bam file}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools view –c smo.bam
\end{Verbatim}
\end{quote}
\begin{itemize}
\item {} 
Can you use samtools to count the number of alignments above a quality score of 20 in your file?

\item {} 
What does the quality score of an alignment indicate?

\item {} 
How many alignments are in the arm.bam above a quality score of 50

\item {} 
What other flags/filtering options are there?

\end{itemize}

{\color{red}\bfseries{}**}Create a bam index **

Use samtools index to create an index of smo.bam

Get help
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools index
\end{Verbatim}
\end{quote}

Note the usage of the index command in the samtools toolbox suite.
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools index smo.bam sm0.bai
\end{Verbatim}
\end{quote}
\begin{itemize}
\item {} 
What does creating an bam index mean?

\item {} 
Why would one want to create a bam index?

\end{itemize}

\textbf{Sorting the sam file}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools sort smo.sam
\end{Verbatim}
\end{quote}
\begin{itemize}
\item {} 
What does –o in the above command do?

\item {} 
Change the command to sort by read names rather than chromosomal locations.

\item {} 
Change the above command so that the sorted reads are outputted to smo.sorted.bam

\end{itemize}

{\color{red}\bfseries{}**}Merging sam files **

Fl1-1.bam and FL1-2.bam are two technical replicates of the one control sample FL1. We would like to merge these two sam files.

Getting help
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools merge
\end{Verbatim}
\end{quote}

Note the usage and run the command
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools merge –h header.sam FL1-merged.bam FL1-1.bam FL1-2.bam
\end{Verbatim}
\end{quote}
\begin{itemize}
\item {} 
What does –h in the above command do?

\item {} 
Does this command work? Why not? (TODO: The files are sorted wrongly). Change the command so that the files can be merged.

\end{itemize}

\textbf{Get summary statistics}

Use \emph{samtools idxstats} to get summary statistics for the aligned file. Use the help.

{\color{red}\bfseries{}**}Create a fasta file index: **
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools faidx dros\_BD5.25.fa
\end{Verbatim}
\end{quote}
\begin{itemize}
\item {} 
What is the benefit of creating an index of a fasta file?

\end{itemize}

{\color{red}\bfseries{}**}Create a pileup **

samtools mpileup is a very useful utility for calling variants in alignment files. Read the help documentation carefully.
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools mpileup –g –l intervalFile.bed –I –D –q 20 -f dros\_BD5.25.fa dnaSeq1.bam dnaSeq2.bam
\end{Verbatim}
\end{quote}

Note: dros\_BD5.25.fa needs to be indexed otherwise the above command will not work


\section{Getting bored? Good to know:}
\label{STEP3_viewManipulateAlignmentData_samtools:getting-bored-good-to-know}
Working with the stream
You can take the output of one command from the “standard stream” and pipe it as input to an(other) samtools command.
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{} samtools view –u dnaSeq1.bam chr1 \textbar{} samtools pileup –cf dros\_BD5.25.fa -
\end{Verbatim}
\end{quote}

Use samtools and awk to count the number of mapped reads in your file.
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{}  samtools idxstats smo.bam \textbar{} awk '\PYGZob{}s+=\$3\PYGZcb{} END \PYGZob{}print s\PYGZcb{}'
\end{Verbatim}
\end{quote}
\begin{itemize}
\item {} 
Is this the same number as with samtools view –c smo.bam

\item {} 
What is the awk command doing in the above?

\end{itemize}


\chapter{Estimating Expression}
\label{STEP4_CUFFLINKS_DE::doc}\label{STEP4_CUFFLINKS_DE:estimating-expression}

\section{Example - Expression in the arm gene knockdown sample}
\label{STEP4_CUFFLINKS_DE:example-expression-in-the-arm-gene-knockdown-sample}
We have used reads sequences obtained from a knockdown of the \emph{arm} and \emph{smo} gene in the drosophila S2 cell line to estimate gene expression. We used tophat to align the reads.  We will now use the output from the tophat step, to look at gene expression in these samples. We will use a program called \textbf{Cufflinks} to do this.

Let's first, take the arm gene knockdown sample and assemble expressed genes and transcripts in the ARM knockdown with CUFFLINKS

To get help
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{cufflinks} \PYG{o}{\PYGZhy{}}\PYG{n}{h}
\end{Verbatim}
\end{quote}

Note the output of them help command
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
Usage: cufflinks [options]* \textless{}aligned\_reads.(sam/bam)\textgreater{}
\end{Verbatim}
\end{quote}

The type abd position of \emph{required} arguments to the cufflinks programs and in angular brackets ``\textless{} \textgreater{}'', while \emph{optional} arguments are in square ``{[}{]}'' brackets.

The output of tophat in step 1 can be found in your home directory at \code{ARM-1\_tophatOutput/accepted\_hits.bam}

The annotation Data is \code{geneRef/dros\_BD5.25.gff}

Run the command:
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
cufflinks -G geneRef/dros.gtf --upper-quartile-norm --compatible-hits-norm -p 2 -o ARM-1\_CufflinksOutput ARM-1\_tophatOutput/accepted\_hits.bam
\end{Verbatim}
\end{quote}


\section{Review Questions:}
\label{STEP4_CUFFLINKS_DE:review-questions}\begin{itemize}
\item {} 
Why assemble transcripts with cufflinks?

\item {} 
What do the options in the cufflinks command do?

\item {} 
What is the gtf file and why do I need it?

\item {} 
what data does the gtf file contain?

\item {} 
what does the last column in the gtf file contain?

\item {} 
what is the ``tss\_id'' tag in the gtf file and what is it used for?

\item {} 
why is both transcript and gene information in the gtf file?

\item {} 
What files does cufflinks output and what do the columns mean?

\end{itemize}

Next, we must estimate the expression in the smo gene knockdown:


\section{Task}
\label{STEP4_CUFFLINKS_DE:task}
Assemble expressed genes and transcripts in the SMO knockdown with CUFFLINKS (similar to above)

The output of tophat in step 1 can be found in your home directory at \code{SMO-1\_tophatOutput/accepted\_hits.bam}


\chapter{Estimating differential Expression}
\label{STEP4_CUFFLINKS_DE:estimating-differential-expression}

\section{Example - Estimating significance of differential expression in the ARM gene knockdown compared to control.}
\label{STEP4_CUFFLINKS_DE:example-estimating-significance-of-differential-expression-in-the-arm-gene-knockdown-compared-to-control}
We have estimated gene expression in SMO and ARM gene knockdowns. We would now like to see if this is expressed differently as one would find in a control. We will now use the output from the cufflinks step, to look at differential gene expression by comparing it to a control. We will use a program called Cuffdiff to do this.

(To save time we have aligned and pre-calculated gene expression in the control samples)

There is 1 replicate of the ARM sample: ARM-1
There are 2 replicates of the CTRL sample: FL1-1, FL2-1

Use the drosophila gtf file \code{geneRef/dros\_BD5.25.gff}

Use the \code{accepted\_hits.bam} output from the tophat alignment

For ctrl these are:
\begin{itemize}
\item {} 
\code{FL1-1\_tophatOutput/accepted\_hits.bam}

\item {} 
\code{FL2-1\_tophatOutput/accepted\_hits.bam}

\end{itemize}

And for ARM these are:\code{ARM-1\_tophatOuput/accepted\_hits.bam}
\begin{quote}

\begin{Verbatim}[commandchars=\\\{\}]
cuffdiff -o ARM\_vs\_CTRL\_diffOut -b genomeRef/dros\_BD5.25.fa -p 8 –L FL1\_ctrl,arm -u geneRef/dros\_BD5.25.gff  FL1-1\_tophatOutput/accepted\_hits.bam, FL2-1\_tophatOutput/accepted\_hits.bam ARM-1\_tophatOuput/accepted\_hits.bam
\end{Verbatim}
\end{quote}
\setbox0\vbox{
\begin{minipage}{0.95\linewidth}
\textbf{Review Questions:}

\medskip

\begin{itemize}
\item {} 
What normalization strategies are available with cuffdiff?

\item {} 
What is the -L option for?

\item {} 
Why are replicates important here?

\item {} 
What are the output files from cuff\_diff?

\end{itemize}
\end{minipage}}
\begin{center}\setlength{\fboxsep}{5pt}\shadowbox{\box0}\end{center}


\section{Use your linux know how}
\label{STEP4_CUFFLINKS_DE:use-your-linux-know-how}\begin{itemize}
\item {} 
How many significantly differentially expressed genes are there?

\item {} 
Can you find the expression of gene ``CG7224''?

\end{itemize}


\section{Task}
\label{STEP4_CUFFLINKS_DE:id1}
Use your know-how to estimate estimate differential expression in the smo knockdown compared to control.

The output can be found at: \code{SMO-1\_tophatOuput/accepted\_hits.bam}

The controls and other input files are as before.


\chapter{Indices and tables}
\label{IntroToCommandLine:indices-and-tables}\begin{itemize}
\item {} 
\emph{genindex}

\item {} 
\emph{modindex}

\item {} 
\emph{search}

\end{itemize}



\renewcommand{\indexname}{Index}
\printindex
\end{document}
